index.js

```diff
@@ -116,33 +97,203 @@ function getConstructorName(errorLike) {
 }
 
 /**
- * ### .getMessage(errorLike)
+ * ### .isCriteria(criteria)
  *
- * Gets the error message from an error.
- * If `err` is a String itself, we return it.
- * If the error has no message, we return an empty string.
+ * Validate that `criteria` is a criteria object by checking that it has an
+ * `.id` property that's strictly equal to the `criteriaMarker` object.
  *
- * @name getMessage
- * @param {Error|String} errorLike
- * @namespace Utils
- * @api public
+ * Note: For performance reasons, this function assumes that `criteria` has
+ * already been verified to be an object.
+ *
+ * @param {Object} criteria
+ * @returns Boolean
  */
+function isCriteria(criteria) {
+  return criteria.id === criteriaMarker;
+}
 
-function getMessage(errorLike) {
-  var msg = '';
-  if (errorLike && errorLike.message) {
-    msg = errorLike.message;
-  } else if (typeof errorLike === 'string') {
-    msg = errorLike;
+/**
+ * ### .createCriteria([errLike[, errMsgMatcher]])
+ *
+ * Return a criteria object which is used internally by `.checkError` and
+ * `.describeExpectedError`. The purpose of this function is to improve
+ * performance when needing to pass the same criteria to both of those other
+ * functions. It allows you to pass `errLike` and/or `errMsgMatcher` a single
+ * time to this function, and then pass the resulting criteria object to both
+ * of those other functions. Doing this ensures that costly type validation is
+ * only performed once per criteria.
+ *
+ * This function allows `errLike` to be omitted and `errMsgMatcher` to be
+ * passed in as the first argument. In other words, this function signature is
+ * also supported: `.createCriteria([errMsgMatcher])`.
+ *
+ * If `errLike` is omitted or is explicitly `null` or `undefined`, then it
+ * defaults to the built-in `Error` constructor.
+ *
+ * See `.checkError` for more info on `errLike` and `errMsgMatcher`.
+ *
+ * @param {Error|ErrorConstructor} [errLike=Error]
+ * @param {String|RegExp} [errMsgMatcher]
+ * @returns Object
+ */
+function createCriteria(errLike, errMsgMatcher) {
+  var errLikeType = type(errLike).toLowerCase();
+  var errMsgMatcherType = type(errMsgMatcher).toLowerCase();
+  // Handle a criteria object being passed by just returning the same object.
+  if (errLikeType === 'object' && isCriteria(errLike)) {
```

Comment:
Seems `isCriteria` could add the additional `typeof errLike === 'object'` check. Additionally - might want to avoid `type-detect` here.